### ISSUE TITLE
[Javascript] special handling of 204 response needed

### ISSUE THREAD

**Issue by tharders** [swagger-api/swagger-codegen#4554]
##### Description

The generated ApiClient.js treats a 204 response with no data as a normal 200 response, where data is expected.

When a request retuns a status 204 with no data in the body, the superagent library will treat this as a success and return an empty object.
The api client will call convertToType with the empty object, which will result in unexpected data.

For array responses this will even result in an error.
When the type parameter is an array convertToType assumes that data is also an array, which it is not.
That will cause the error "data.map is not a function"

##### Swagger-codegen version

Using http://editor.swagger.io

##### Swagger declaration file content or url

```yaml
swagger: '2.0'
info:
  version: "0.0.1"
  title: Test API
produces:
  - application/json
paths:
  /test:
   get:
      responses:
        200:
          description: Success
          schema:
            type: array
            description: array of results
            items:
              type: string
```

##### Command line used for generation

http://editor.swagger.io

##### Steps to reproduce

Generate a javascript client with the example code and let the server return 204 without any body. 
Then calling the api should reproduce the problem.

##### Related issues

I saw that the 204 was an issue with other languages.
Maybe it should be testest with the all languages if the 204 is handled correctly.

##### Suggest a Fix
I fixed this by checking for status 204 in the deserialize method of the ApiClient.js and returning null in this case.
```javascript
    exports.prototype.deserialize = function deserialize(response, returnType) {
        if (response == null || returnType == null) {
            return null;
        }
        if (response.status == 204) {
            return null;
        }
```

I also added a try catch around the call to deserialize to avoid unhandled exceptions like the  "data.map is not a function"
```javascript
                if (!error) {
                    try {
                        data = _this.deserialize(response, returnType);
                    } catch (err) {
                        error = err;
                    }
                }
```

**Comment by wing328** [comment]
@tharders thanks for reporting the issue.

May I know if you've time to submit a PR for the enhancement?

Personally I would combine the if statements in 1 line:
```
        if (response == null || returnType == null || response.status == 204) {
```

**Comment by tharders** [comment]
Hi @wing328,
I have only looked at the  generated code and not at the code generator itself.
So I don't think it makes sense that I dig into this, since there is alsso not that much time I can spare.

**Comment by wing328** [comment]
@tharders that's ok. More than 450+ developers have made contributions to this project so I would say the process for contributing back is very mature and straightforward, and won't take you a lot of time.

Anyway, if I make the change to the master, would you have time to do a test ?

**Comment by tharders** [comment]
@wing328 you say its easy, so I gave it all a try.
I followed the steps in the contribution documentation and here is the pull request:
https://github.com/swagger-api/swagger-codegen/pull/4604
It was a little bit confusing that the windows script to generate the petstore client was not updated in accordance with the shell script, but I also changed that.
I hope I did't miss anything.

**Comment by wing328** [comment]
Closed via #4604
